specs/ContextMenuRequested.md

```diff
@@ -0,0 +1,764 @@
+
+        /// Specifies a separator menu item kind. `ContextMenuItem` objects of this kind
+        /// are used to signal a visual separator with no functionality.
+        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SEPARATOR,
+
+        /// Specifies a submenu menu item kind. `ContextMenuItem` objects of this kind will contain
+        /// a `ContextMenuItemCollection` of its children `ContextMenuItem` objects.
+        COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND_SUBMENU,
+    } COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND;
+
+    /// Represents a context menu item of a context menu displayed by WebView.
+    [uuid(7aed49e3-a93f-497a-811c-749c6b6b6c65), object, pointer_default(unique)]
+    interface ICoreWebView2ContextMenuItem : IUnknown
+    {
+        /// Gets the localized label for the `ContextMenuItem`.
+        [propget] HRESULT Label([out, retval] LPWSTR* value);
+
+        /// Gets the unlocalized name for the `ContextMenuItem`. Use this to distinguish items with 
+        /// same Descriptor such as Extension items and items with Other as descriptor.
+        [propget] HRESULT Name([out, retval] LPWSTR* value);
+
+        /// Gets the descriptor for the `ContextMenuItem`.
+        [propget] HRESULT Descriptor([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_DESCRIPTOR* value);
+
+        /// Gets the Command ID for the `ContextMenuItem`. Use this to report the `SelectedCommandId` in `ContextMenuRequested` event.
+        [propget] HRESULT CommandId([out, retval] INT32* value);
+
+        /// Gets the keyboard shortcut for this ContextMenuItem. It will be the empty
+        /// string if there is no keyboard shortcut.
+        /// This is text intended to be displayed to the end user to show the keyboard shortcut. 
+        /// For example this property is `Ctrl+Shift+I` for the "Inspect" `ContextMenuItem`.
+        [propget] HRESULT ShortcutKeyDescription([out, retval] LPWSTR* value);
+
+        /// Gets the Icon for the `ContextMenuItem` in PNG format in the form of an IStream.
+        [propget] HRESULT Icon([out, retval] IStream** value);
+
+        /// Gets the `ContextMenuItem` kind.
+        [propget] HRESULT Kind([out, retval] COREWEBVIEW2_CONTEXT_MENU_ITEM_KIND* value);
+
+        /// Gets the enabled property of the `ContextMenuItem`.
+        [propget] HRESULT IsEnabled([out, retval] BOOL* value);
+
+        /// Gets the checked property of the `ContextMenuItem`, used if the kind is Checkbox or Radio.
+        [propget] HRESULT IsChecked([out, retval] BOOL* value);
+
+        /// Gets the list of children menu items through a `ContextMenuItemCollection` 
+        /// if the kind is Submenu. If the kind is not submenu, will return null.
+        [propget] HRESULT Children([out, retval] ICoreWebView2StagingContextMenuItemCollection** value);
+
+        /// Add an event handler for the `CustomItemSelected` event.
+        /// `CustomItemSelected` event is raised when the user selects this `ContextMenuItem`.
+        HRESULT add_CustomItemSelected(
```

Comment:
What happens if I subscribe to the CustomItemSelected event on a system-provided MenuItem? Do I get called when the user picks that menu item?

Reply:
Talking with @Hosamtag CustomItemSelected is only raised for end developer created context menu items and won't be raised by system-provided context menu items. 

To me this seems a little weird, but the alternative I can think of is to rename the event to just ItemSelected, raise it for system-provided context menu items as well as custom, add a Cancel boolean property to the event args to control whether the system provided functionality is also invoked or not, and also a GetDeferral. This seems like a lot of extra work for something not required for the scenario. 

@oldnewthing, @MikeHillberg, any thoughts on this?

Reply:
I'm okay with CustomItemSelected being raised only for custom items. That's why we put "Custom" in the name of the event: It applies only to custom items.

If we were willing to be all COM interface-y, we could have a separate interface `ICoreWebView2CustomContextMenuItem`, and that interface has the `CustomItemSelected` event. System-provided context menu items do not implement that interface, so it becomes physically impossible to register a listener on them. (No surprise that an event that doesn't exist is never raised.)

If you wanted to go all crazy, you could go all in on the COM interface pattern and have

* ICoreWebView2ContextMenuItem: Kind, Name, CommandId
* ICoreWebView2NormalContextMenuItem: ShortcutKeyDescription, Icon, Label, IsEnabled
* ICoreWebView2CheckableContextMenuItem: IsChecked
* ICoreWebView2SubmenuContextMenuItem: Children
* ICoreWebView2CustomContextMenuItem: CustomItemSelected

| Scenario | Implemented interfaces |
|-|-|
| Separator | ICoreWebView2ContextMenuItem only |
| Normal | ICoreWebView2ContextMenuItem + ICoreWebView2NormalContextMenuItem |
| CheckBox | ICoreWebView2ContextMenuItem + ICoreWebView2NormalContextMenuItem + ICoreWebView2CheckableContextMenuItem |
| Radio | ICoreWebView2ContextMenuItem + ICoreWebView2NormalContextMenuItem + ICoreWebView2CheckableContextMenuItem |
| Submenu | ICoreWebView2ContextMenuItem + ICoreWebView2NormalContextMenuItem + ICoreWebView2SubmenuContextMenuItem |

Custom versions of the above would also add ICoreWebView2CustomContextMenuItem.

(I think this goes a bit too far though.)